libstuff/SDeburr.cpp

```diff
@@ -0,0 +1,137 @@
+            result.append(reinterpret_cast<const char*>(input_bytes + i), 1);
+            i++;
+            continue;
+        }
+
+        // Check for invalid UTF-8. If invalid, skip the current byte and continue.
+        // Cases we check for:
+        // - A continuation byte (1 leading one). This should be handled only as part of a multi-byte sequence.
+        // - More than 4 leading ones
+        // - Not enough bytes to form a valid sequence
+        if (numLeadingOnes == 1 || numLeadingOnes > 4 || (i + numLeadingOnes) > len) {
+            i++;
+            continue;
+        }
+
+        /*
+         * This is a valid 2, 3, or 4 byte UTF-8 sequence, so next we'll decode the full character.
+         *
+         * The way this works is best explained by example.
+         *
+         * The hex codepoint for ñ is U+00F1, which is represented in UTF-8 as a 2-byte sequence: 11000011 10110001
+         *
+         *   - The first byte has 2 leading ones, which tells us it's a 2-byte sequence.
+         *   - The next byte has 1 leading one, which tells us it's a continuation byte (a trailing part of a multi-byte sequence).
+         *
+         * To get the codepoint (the single number representing the character), we take all the bits from the sequence that aren't "signaling" other information and concat them together.
+         *
+         *   1. Remove the prefix bits (110) from the first byte. That gives us 00011.
+         *   2. Remove the prefix bits (10) from the second byte. That gives us 110001.
+         *   3. Concat these together, and the codepoint for ñ is 00011110001 (which, sure enough, is the binary representation of U+00F1).
+         */
+        size_t start = i;
+
+        // Extract data bits from leading byte
+        // The number of prefix bits is 7 - numLeadingOnes, because the leading ones signaling the length of the sequence are always terminated by a 0.
+        // For example, two leading ones for a 2-byte sequence results in a prefix of 110 in the first byte of the sequence.
+        uint32_t codepoint = input_bytes[i] & ((1 << (7 - numLeadingOnes)) - 1);
+        i++;
+
+        // Process continuation bytes. 
+        for (int j = 1; j < numLeadingOnes; j++) {
+            if (countl_one(input_bytes[i]) != 1) {
+                // Invalid continuation byte, stop processing this sequence
+                break;
+            }
+
+            // Shift the existing codepoint over 6 positions to make room for the new bits
+            codepoint <<= 6;
+
+            // Extract lower 6 bits from continuation and concat them to the existing codepoint
+            codepoint = codepoint | (input_bytes[i] & 0b00111111);
+            i++;
+        }
+
+        // Accent marks by themselves (like ´ ` ^) → skip them
+        if (codepoint >= 0x0300 && codepoint <= 0x036F) {
+            continue;
+        }
+
+        // Map the unicode codepoint to ascii
+        const char* mapped = unicodeToAscii(codepoint);
```

Comment:
I think it would make more sense for this to return `char` instead of a `char *`. It can still return 0 as the not found value.

Reply:
Well it returns `char*` because some unicode characters are represented as more than one ascii characters (`œ` becomes `oe`, for example).

I could make it return a C++ string though, which I think would simplify things.

Reply:
Hmmm the table is constexpr so you can't use a non-transient `std::string` in there. We could convert it to a string by just using the `string()` constructor, but that seems like unnecessary overhead vs just appending the `char*` to the output string directly.